### ISSUE TITLE
prediction based on .sdf files

### ISSUE THREAD

**Issue by akourm910e** [deepchem/deepchem#1294]
**Description:**
Hello, I am new to deepchem and deeplearning.
I would like to understand something, I train a model from the membrane_permeability example, for this dataset we load the sdf data and the associated csv file. And we train him.
But how does one make prediction on sdf files without the associated csv?
Is it possible ?
Thank you

- DeepChem Version: 2.0
- Platform: ubuntu16.04

```
Your code sample here
```

**Comment by rbharath** [comment]
Yes, this is possible. Use `SDFLoader` to load the molecules from the SDF file into a `Dataset` object and make predictions.

**Comment by akourm910e** [comment]
Thank you for response, 
when i try to load SDF files, it ask me csv file, but i have just the .sdf for my data. 
Should i create an empty csv file for my sdf data? 

To train model there is no problem, i want to make prediction on my data.

**Comment by akourm910e** [comment]
def predict(): 

    featurizer_func = dc.feat.ConvMolFeaturizer()
    current_dir = os.path.dirname(os.path.realpath(__file__))
    test_dataset = os.path.join(current_dir,
                              "data/test_data.sdf")
    print("About to featurize test dataset.")
    tasks = sorted(['LogP(RRCK)'])
    loader = dc.data.SDFLoader(
                            tasks=tasks, clean_mols=True, featurizer=featurizer_func)
    dataset = loader.featurize(test_dataset)   
    
    model2 = model.load_from_dir('models')
    
    test_predict = model2.predict(dataset)

**Comment by akourm910e** [comment]
def predict():

featurizer_func = dc.feat.ConvMolFeaturizer()
current_dir = os.path.dirname(os.path.realpath(__file__))
test_dataset = os.path.join(current_dir,
                          "data/test_data.sdf")
print("About to featurize test dataset.")
tasks = sorted(['LogP(RRCK)'])
loader = dc.data.SDFLoader(
                        tasks=tasks, clean_mols=True, featurizer=featurizer_func)
dataset = loader.featurize(test_dataset)   

model2 = model.load_from_dir('models')

test_predict = model2.predict(dataset)

**Comment by akourm910e** [comment]
def predict(): 

    featurizer_func = dc.feat.ConvMolFeaturizer()
    current_dir = os.path.dirname(os.path.realpath(__file__))
    test_dataset = os.path.join(current_dir,
                              "data/test_data.sdf")
    print("About to featurize test dataset.")
    tasks = sorted(['LogP(RRCK)'])
    loader = dc.data.SDFLoader(
                            tasks=tasks, clean_mols=True, featurizer=featurizer_func)
    dataset = loader.featurize(test_dataset)   
 
    model2 = model.load_from_dir('models')
    
    test_predict = model2.predict(dataset)

**Comment by rbharath** [comment]
This might be an infelicity in our current SDF file handling. 

@lilleswing Do we require a csv file for making predictions on SDFs?

**Comment by lilleswing** [comment]
we do -- it is a legacy thing we should get rid of

@akourm910e you can always make the dataset yourself.

```python
from rdkit import Chem
import numpy as np
import deepchem as dc


def load_sdf(fname, tasks=list(), featurizer=None):
  """
  Parameters
  ----------
  fname: str
    filename of sdf to load
  tasks: list of str
    tasks from the sdf file to load
  Returns
  -------
  """
  suppl = Chem.SDMolSupplier(fname)
  mols = [x for x in suppl]
  if featurizer is None:
    X = np.array(mols)
  else:
    X = featurizer.featurize(mols)
  if len(tasks) == 0:
    y = np.zeros((len(mols), 1))
    return dc.data.NumpyDataset(X, y)

  y = np.zeros((len(mols), len(tasks)))
  w = np.zeros((len(mols), len(tasks)))
  for i, mol in enumerate(mols):
    for j, task in enumerate(tasks):
      try:
        y[i][j] = float(mol.GetProp(task))
        w[i][j] = 1
      except:
        w[i][j] = 0

  return dc.data.NumpyDataset(X, y, w)


if __name__ == "__main__":
  ds = load_sdf("datasets/membrane_permeability.sdf",
                ["LogP(RRCK)"],
                featurizer=dc.feat.ConvMolFeaturizer())
  model = dc.models.GraphConvModel(1,mode='regression')
  model.fit(ds)
  model.predict(ds)
```

**Comment by rbharath** [comment]
@lilleswing This looks like a really useful utility function. Should we add this somewhere to the main repo?

**Comment by rbharath** [comment]
Agree we should swap to an in-memory model for base case. Few datasets are really big enough to justify the on-disk default

**Comment by BJWiley233** [comment]
Just looking at example loading `gdb7.sdf` from AWS for tutorial "Tutorial Part 21: Exploring Quantum Chemistry with GDB".  The "atomization_energy" is used in the tutorial but in the csv file under AWS had the header column "u0_atom".  I was getting a key error from pandas but thought it was an issue with loading.  Might be nice to add in doc string that headers in csv files must match task names in the loaders.

**Comment by rbharath** [comment]
@BJWiley233 To clarify, were you seeing an error in the tutorial itself? If so we should certainly fix that

**Comment by BJWiley233** [comment]
No no error.  You just don’t see the header of the csv file in tutorial or really very often when downloading datasets that come with a csv and so if you have a different task name then the header it won’t create the dataset.

**Comment by peastman** [comment]
As far as I can tell, the fact that the SDF loader looks for a CSV file in the same location with the same name is totally undocumented.  We really need to improve the docs for it.
